test/IRGen/loadable_by_address_issue73018.swift

```diff
@@ -0,0 +1,44 @@
+// REQUIRES: differentiable_programming
+// RUN: %target-swift-frontend -emit-ir -verify %s
+
+// This used to trigger an assertion due to LoadableByAddress not doing proper mapping of
+// switch_enum arguments during rewriting
+
+import Foundation; import _Differentiation
+struct O: Differentiable {var a: B.G; var b: Array<SIMD2<Float>>; var c: B.M; var d: B.M}
```

Comment:
Is there a smaller reduction that doesn’t use Differentiable?

Reply:
I do not have one sadly. Differentiation tends to create `switch_enum`'s with large payloads having tuples with closures inside. So, this exposes all kinds of corner cases in `LoadableByAddress`.

The type in question here is:
```
$(@callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@guaranteed C<SIMD2<Float>>) -> @owned Array<SIMD2<Float>>.DifferentiableView, @callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@guaranteed B.M.TangentVector) -> @owned B.M.TangentVector, @callee_guaranteed (@guaranteed B.M.TangentVector) -> @owned B.M.TangentVector))
```

that is mapped to:
```
$(@callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@guaranteed C<SIMD2<Float>>) -> @owned Array<SIMD2<Float>>.DifferentiableView, @callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@guaranteed C<Float>) -> @owned Array<Float>.DifferentiableView, @callee_guaranteed (@in_guaranteed B.M.TangentVector) -> @out B.M.TangentVector, @callee_guaranteed (@in_guaranteed B.M.TangentVector) -> @out B.M.TangentVector)
```

And everything seems to depend on the order of previous transformations. I cannot remove any single line in the reproducer to still trigger an assertion, looks like we need to hit some thresholds as in:
```c++
  bool isLargeLoadableType(SILType ty) {
    if (ty.isAddress() || ty.isClassOrClassMetatype()) {
      return false;
    }

    auto canType = ty.getASTType();
    if (canType->hasTypeParameter()) {
      assert(genEnv && "Expected a GenericEnv");
      canType = genEnv->mapTypeIntoContext(canType)->getCanonicalType();
    }

    if (canType.getAnyGeneric() || isa<TupleType>(canType)) {
      assert(ty.isObject() &&
             "Expected only two categories: address and object");
      assert(!canType->hasTypeParameter());
      const TypeInfo &TI = irgenModule->getTypeInfoForLowered(canType);
      auto &nativeSchemaOrigParam = TI.nativeParameterValueSchema(*irgenModule);
      if (nativeSchemaOrigParam.size() > 15)
        return true;
      auto explosionSchema = TI.getSchema();
      if (explosionSchema.size() > 15)
        return true;
    }
    return false;
  }
```

Reply:
The thresholds in that function are just a property of the stored property layout of the type and it should be possible to hit them by declaring, eg, a struct with a tuple containing 100 elements.